src/libsyntax/parse/parser.rs

```diff
@@ -515,6 +515,12 @@ impl<'a> Parser<'a> {
         }
     }
 
+    pub fn error_if_typename_is_catch(&mut self, ident: ast::Ident) {
```

Comment:
`error_if_typename_is_catch` seems unnecessary and it breaks more cases than strictly needed. It's not an error to simply create a struct named `catch`. Something like
```
struct catch;
```
can be created without `catch {}` expression and even
```
struct catch { field: u8 }
```
can be created like this
```
type Alias = catch;
let c = Alias { field: 0 };
```

*Unless* we want to make `catch` a keyword, then it needs to warned against everywhere (all idents) and not only in item names.